File: src/types/system_control.ts
export interface NetworkAdapter {
  /** GUID identifying this adapter — pass this (never `name`) to changeDns. */
  adapter_id: string;
  name: string;
  description: string;
  adapter_type: "Ethernet" | "Wi-Fi" | "VPN" | "Loopback" | "Other";
  status: "Up" | "Down" | "Unknown";
  mac_address: string | null;
  ipv4_addresses: string[];
  ipv6_addresses: string[];
  gateway: string | null;
  dns_servers: string[];
  dhcp_enabled: boolean | null;
  link_speed_mbps: number | null;
}

export type ServiceStatus =
  | "RUNNING"
  | "STOPPED"
  | "PAUSED"
  | "START_PENDING"
  | "STOP_PENDING"
  | "UNKNOWN";

export type StartupType =
  | "AUTOMATIC"
  | "AUTOMATIC_DELAYED"
  | "MANUAL"
  | "DISABLED"
  | "UNKNOWN";

export interface ServiceInfo {
  name: string;
  display_name: string;
  status: ServiceStatus;
  startup_type: StartupType;
  executable: string | null;
  description: string | null;
  protected: boolean;
}

export type ServiceAction = "START" | "STOP" | "RESTART";

export interface ServiceActionRequest {
  service_name: string;
  action: ServiceAction;
}

export interface ChangeStartupTypeRequest {
  service_name: string;
  startup_type: StartupType;
}

export type DnsMode = "DHCP" | "STATIC";

export interface DnsSettingsRequest {
  adapter_id: string;
  mode: DnsMode;
  /** Required when mode is STATIC. */
  primary_dns: string | null;
  secondary_dns: string | null;
}

File: src-tauri/src/models/system_control.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NetworkAdapter {
    /// The adapter's GUID (e.g. `{4D36E972-...}`), from `GetAdaptersAddresses`'
    /// `AdapterName` field. This — never `name` — is what identifies the
    /// adapter to `commands::dns::change_dns`: friendly names can repeat
    /// or change, the GUID doesn't.
    pub adapter_id: String,
    pub name: String,
    pub description: String,
    pub adapter_type: String, // "Ethernet" | "Wi-Fi" | "VPN" | "Loopback" | "Other"
    pub status: String,       // "Up" | "Down" | "Unknown"
    pub mac_address: Option<String>,
    pub ipv4_addresses: Vec<String>,
    pub ipv6_addresses: Vec<String>,
    pub gateway: Option<String>,
    pub dns_servers: Vec<String>,
    pub dhcp_enabled: Option<bool>,
    pub link_speed_mbps: Option<u64>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "UPPERCASE")]
pub enum ServiceStatus {
    Running,
    Stopped,
    Paused,
    StartPending,
    StopPending,
    Unknown,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum StartupType {
    Automatic,
    AutomaticDelayed,
    Manual,
    Disabled,
    Unknown,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ServiceInfo {
    pub name: String,
    pub display_name: String,
    pub status: ServiceStatus,
    pub startup_type: StartupType,
    pub executable: Option<String>,
    pub description: Option<String>,
    /// Services Windows itself treats as protected (cannot safely be
    /// stopped without risking system stability). The UI must show an
    /// extra-strength confirmation for these — see SECURITY.md.
    pub protected: bool,
}

#[derive(Debug, Clone, Copy, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ServiceAction {
    Start,
    Stop,
    Restart,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ServiceActionRequest {
    pub service_name: String,
    pub action: ServiceAction,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ChangeStartupTypeRequest {
    pub service_name: String,
    pub startup_type: StartupType,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum DnsMode {
    Dhcp,
    Static,
}

/// Typed request to change one adapter's DNS configuration. Always
/// re-validated server-side (see `commands::dns::change_dns`) — the
/// frontend's own IPv4 checks are just for fast feedback, never trusted.
#[derive(Debug, Clone, Deserialize)]
pub struct DnsSettingsRequest {
    /// Must be a `NetworkAdapter::adapter_id`, not a display name.
    pub adapter_id: String,
    pub mode: DnsMode,
    /// Required when `mode` is `STATIC`; ignored for `DHCP`.
    pub primary_dns: Option<String>,
    /// Optional even in `STATIC` mode.
    pub secondary_dns: Option<String>,
}
